### PR TITLE
Trilinos Dependencies For Windows v2

### PR DESCRIPTION
**Pull request by jpopelar**
For porting of packages that Trilinos requires to build. A far cleaner and more recent version of #33801. Will include boost and py-setuptools to start with; will need to consult with others about makefile packages and whether it is fine to port them over to CMake (esp. with openblas)

### CODE REVIEW

File: var/spack/repos/builtin/packages/boost/package.py
Comment (hainest) [inline var/spack/repos/builtin/packages/boost/package.py]: Was this happening with just the classic intel compiler or with oneapi, as well?

File: var/spack/repos/builtin/packages/boost/package.py
Comment (hainest) [inline var/spack/repos/builtin/packages/boost/package.py]: Is this some debug logging to-be-removed?

File: var/spack/repos/builtin/packages/boost/package.py
Comment (hainest) [inline var/spack/repos/builtin/packages/boost/package.py]: The `"--prefix=%s"` wasn't propagated to the `else` clause below. Was that intentional?

File: var/spack/repos/builtin/packages/boost/package.py
Comment (hainest) [inline var/spack/repos/builtin/packages/boost/package.py]: Could we merge this with line 717 so we only do the check once?

File: var/spack/repos/builtin/packages/pcre/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/pcre/package.py]: Could you refactor this to be a multi-build-system package? i.e. inherit `CMakePackage` and `AutotoolsPackage` etc.

File: var/spack/repos/builtin/packages/boost/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/boost/package.py]:
```
# We also use a hack right now to get around the fact we don't have
# a wrapper in spack/env for cl.exe
```

I think it would be clearer to rephrase this as:

```
There is no spack wrapper for cl.exe, so skip this on Windows
```

File: var/spack/repos/builtin/packages/boost/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/boost/package.py]: This appears to undo https://github.com/spack/spack/commit/38592b9e4f974254c18ce82e8b7ccca65a07539e - have you confirmed there was an issue with that change on Linux?

File: var/spack/repos/builtin/packages/boost/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/boost/package.py]: Avoid using the term "again": it makes sense within the context of the PR, but once the changes get merged into the file, it provides no real context.

File: var/spack/repos/builtin/packages/boost/package.py
Comment (scheibelp) [reply var/spack/repos/builtin/packages/boost/package.py]: (I might be saying the same thing as the above user but wanted to make sure to ask) With this change, I don't see `prefix` being specified anywhere: how is boost installing to the appropriate directory?

**Review by scheibelp — CHANGES_REQUESTED** [review]
Initial requests

**Comment by jpopelar** [comment]
@scheibelp PTAL when convenient. Will fix style issues momentarily and also email you a text doc with some lingering trilinos issues (mostly related to the shared and mpi variants)

File: var/spack/repos/builtin/packages/boost/package.py
Comment (scheibelp) [reply var/spack/repos/builtin/packages/boost/package.py]: This is still true, but will be untrue pending https://github.com/spack/spack/pull/38495 (which is not merged yet and probably won't be merged until after this, but I note it here for reference).

File: var/spack/repos/builtin/packages/openblas/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/openblas/package.py:24]: (no change needed) Note this attribute is just for finding external instances. We might want to automatically strip `lib` on Windows in `detection/path.py` (doesn't have to happen in this PR)

File: var/spack/repos/builtin/packages/openblas/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/openblas/package.py]: Note: `makefile.py` already declares a conflict on Windows when `build_system=makefile`, so this should be unnecessary (also, this doesn't technically force using the Cmake build system, but that's a more-abstract reason).

File: var/spack/repos/builtin/packages/openblas/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/openblas/package.py]: This class appears unnecessary.

File: var/spack/repos/builtin/packages/openblas/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/openblas/package.py]: Note: if you had to specify `libraries = [libopenblas, openblas]` above, I'd be surprised if you didn't have to do it here.

File: var/spack/repos/builtin/packages/openblas/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/openblas/package.py:515]: Because you set CMake as the default build system everywhere (including Linux), we'd be particularly interested in supporting all the variants and build functionality we supported with Makefiles, which includes the variants:

* shared
* fortran
* locking
* ilp64
* consistent_fpcsr
* bignuma

I’m guessing you’d also want to promote _microarch_target_args from a Makefile-specific method (although `GENERIC` is the sensible choice on Windows for now, we'd want to be specific when building with CMake on  other systems), unless you see CMake handling this itself.

We could keep the CMake build stripped-down, but I'd say in that case it should only be available on Windows.

File: var/spack/repos/builtin/packages/superlu/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/superlu/package.py]: This should be `conditional(generic, when=@:4)` (this inherits `Package` rather than `AutotoolsPackage`).

File: var/spack/repos/builtin/packages/openblas/ifort-msvc.patch
Comment (scheibelp) [inline var/spack/repos/builtin/packages/openblas/ifort-msvc.patch:10]: (question) Do you know if there are any pending PRs in Boost itself for this?

File: var/spack/repos/builtin/packages/boost/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/boost/package.py]: Don't clear this at the end, do something like:

```
if spec.satisfies(platform=windows)
  b2_options = []
else:
  all the logic to set b2 options
```

I'll need to read this carefully to know if there's any options in `b2_options` that are not specified in `bootstrap_options`, and if so my concern would be that we aren't customizing the build as we would want.

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/trilinos/package.py]: Customize each message to the variant, e.g.:

```
Only static builds are supported on Windows currently
```

somewhat repetitious, but that allows us to address each of these in potentially separate PRs without the message becoming wrong.

**Review by scheibelp — CHANGES_REQUESTED** [review]
Additional requests

File: var/spack/repos/builtin/packages/openblas/ifort-msvc.patch
Comment (jpopelar) [reply var/spack/repos/builtin/packages/openblas/ifort-msvc.patch:10]: Not that I am aware of, no, but I can verify

File: var/spack/repos/builtin/packages/boost/package.py
Comment (jpopelar) [reply var/spack/repos/builtin/packages/boost/package.py]: For sure. Either way, cleaned up for posterity in next commit

**Comment by jpopelar** [comment]
@scheibelp Looks like prechecks are passing. PTAL whenever convenient and we'll get this reviewed and merged asap

File: var/spack/repos/builtin/packages/boost/package.py
Comment (scheibelp) [reply var/spack/repos/builtin/packages/boost/package.py]: Bump on this question.

File: var/spack/repos/builtin/packages/boost/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/boost/package.py]: Originally you removed the `--prefix=` from install so it was set nowhere (i.e. https://github.com/spack/spack/pull/34622#discussion_r1063894135). Now you restored that and set it here, so it's set twice.

File: var/spack/repos/builtin/packages/boost/package.py
Comment (scheibelp) [reply var/spack/repos/builtin/packages/boost/package.py]: Bump: you can move 

```
            path_to_config = "--user-config=%s" % os.path.join(
                self.stage.source_path, "user-config.jam"
            )
```

like

```
        if not is_windows:
            path_to_config = "--user-config=%s" % os.path.join(
                self.stage.source_path, "user-config.jam"
            )
            b2_options = ["-j", "%s" % jobs]
            b2_options.append(path_to_config)
```

File: var/spack/repos/builtin/packages/boost/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/boost/package.py]: How does this install succeed on windows, as below we have

```
        for threading_opt in threading_opts:
            b2("install", "threading=%s" % threading_opt, *b2_options)
```

(including on Windows) but the `threading_opts` variable is not defined on Windows.

File: var/spack/repos/builtin/packages/openblas/ifort-msvc.patch
Comment (scheibelp) [reply var/spack/repos/builtin/packages/openblas/ifort-msvc.patch:10]: Checking in on this

File: var/spack/repos/builtin/packages/openblas/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/openblas/package.py:96]: It's great that you linked this, but does this mean that only certain versions need to apply this patch (e.g. it looks like that PR was merged in 2020, so perhaps all versions after that don't need this patch)?

File: var/spack/repos/builtin/packages/openblas/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/openblas/package.py]: Drop the "as in netlib-lapack" (this won't provide much context to people).

File: var/spack/repos/builtin/packages/openblas/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/openblas/package.py]: Note that the first argument of `find_libraries` allows a string, but only if that string represents a single library. If you want to look for multiple, you need to break it into a list of strings.

Also, you'd need to individually append the symbol suffix anyway. That being said, I don't see any `symbol_suffix` variant in `boost` right now and I don't see this diff adding one. Without the `symbol_suffix` logic, this matches the behavior of `Spec._libs_default_handler`.

File: var/spack/repos/builtin/packages/openblas/package.py
Comment (scheibelp) [reply var/spack/repos/builtin/packages/openblas/package.py:515]: You may have added a few of these options since, but I still don't see handling of `ilp64`, `consistent_fpcsr`, `locking`, or `bignuma` (so I'd still be inclined not to make `CMake` the default on systems other than Windows).

File: var/spack/repos/builtin/packages/superlu/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/superlu/package.py]: Do you know if cmake builds were available for versions before 5? I would guess they weren't in which case this should be replaced with a separate `conflicts(platform=windows, when=@:4)`.

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/trilinos/package.py]: Can you remove this comment (it's not wrong, but I don't think it's necessary)?

**Review by scheibelp — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/openblas/package.py
Comment (jpopelar) [reply var/spack/repos/builtin/packages/openblas/package.py:96]: Decent point! I can try building after my next commit here and let you know

File: var/spack/repos/builtin/packages/openblas/ifort-msvc.patch
Comment (jpopelar) [reply var/spack/repos/builtin/packages/openblas/ifort-msvc.patch:10]: Just checked, the only recent PR that even mentions MSVC is closed and only touches something in the command prompt. Hence I believe this patch is still required

File: var/spack/repos/builtin/packages/boost/package.py
Comment (jpopelar) [reply var/spack/repos/builtin/packages/boost/package.py]: I'm honestly not completely sure; regardless I've insulated it better now

File: var/spack/repos/builtin/packages/boost/package.py
Comment (jpopelar) [reply var/spack/repos/builtin/packages/boost/package.py]: I had this on oneAPI as well I believe, though I'd need to review what happened

File: var/spack/repos/builtin/packages/superlu/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/superlu/package.py:47]: This `build_system` declaration is still required. If you restore it as it was, in combination with the conflict you added just below, the desirable result should be achieved in all circumstances.

Also I realized that replacing the `conflict` with:

```
require(build_system=cmake, when="platform=windows")
```

would be slightly more robust: for now since the support ranges do not overlap they are equivalent, but if they ever did for some reason, I assume we would want to force building with CMake on Windows.

File: var/spack/repos/builtin/packages/boost/package.py
Comment (scheibelp) [reply var/spack/repos/builtin/packages/boost/package.py]: You no longer need to set a default value for `path_to_config` here.

File: var/spack/repos/builtin/packages/openblas/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/openblas/package.py:279]: I'm already seeing an existing implementation of libs/headers in openblas, and I think it would cover what you need. Note that other implementation does not search for "openblas.lib", but `find_libraries` actually expects to do all the suffix appending itself, so in fact we should not be passing "openblas.lib" to it.

File: var/spack/repos/builtin/packages/boost/package.py
Comment (scheibelp) [reply var/spack/repos/builtin/packages/boost/package.py]: I'll also note that `determine_b2_options` determines more than just threading options. Are all of the options handled by that function not relevant on Windows? I see for example the following variants handled:

* `+debug`
* `+icu`

And other important-seeming things not determined by variants (e.g. c++ standard).

My guess would be that you'd want to do this everywhere, but keep the below if (where on Windows you don't invoke `b2` with `threading=`).

**Review by scheibelp — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/superlu/package.py
Comment (jpopelar) [reply var/spack/repos/builtin/packages/superlu/package.py:47]: Are we sure that syntax is correct? I currently have

```
    build_system(
        conditional("cmake", when="@5:"), conditional("generic", when="@:4"), default="cmake"
    )

    requires(build_system="cmake", when="platform=windows")
```

CI then tells me that I can't use build_system in a requires() directive

File: var/spack/repos/builtin/packages/boost/package.py
Comment (scheibelp) [reply var/spack/repos/builtin/packages/boost/package.py]: I'm looking closer at https://github.com/spack/spack/commit/38592b9e4f974254c18ce82e8b7ccca65a07539e and I think this is also on account of a merge conflict: I think you can make the if statement just:

```
if not spec.satisfies("platform=windows"):
```

File: var/spack/repos/builtin/packages/superlu/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/superlu/package.py]: This needs to be `requires("build_system=cmake", when=...)`

File: var/spack/repos/builtin/packages/openblas/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/openblas/package.py]: This should be:

```
name = ["libopenblas", "openblas"]
```

(`find_libraries` just wants the name, and will append `.lib` itself).

File: var/spack/repos/builtin/packages/boost/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/boost/package.py]: I'm assuming the idea is you don't want to set `toolset=` if we're on windows, in that case, this `if` would need to be:

`if not (spec.satisfies("@:1.75 %intel") and not spec.satisfies("platform=windows")):`

**Review by scheibelp — CHANGES_REQUESTED** [review]
I have a few minor requests. Once those are resolved, can you remove the draft status? The CI will start building some of these packages.

File: var/spack/repos/builtin/packages/superlu/package.py
Comment (jpopelar) [reply var/spack/repos/builtin/packages/superlu/package.py]: So I tried that earlier but it broke prechecks. Apparently you can't have `build_system` stuff inside a requires block. I figured this would be the closest to what we want? Is there another way to do this?

File: var/spack/repos/builtin/packages/boost/package.py
Comment (jpopelar) [reply var/spack/repos/builtin/packages/boost/package.py]: Sorry, I deMorgan's lawed that one. Your method is definitely clearer

File: var/spack/repos/builtin/packages/superlu/package.py
Comment (jpopelar) [reply var/spack/repos/builtin/packages/superlu/package.py]: @scheibelp Just making sure you saw this

Comment (jpopelar) [reply var/spack/repos/builtin/packages/superlu/package.py]: @scheibelp Bump. I'll rebase and merge as soon as I get an answer on this

File: var/spack/repos/builtin/packages/boost/package.py
Comment (scheibelp) [reply var/spack/repos/builtin/packages/boost/package.py]: I misread what you originally had, and I should have been OK with it. Thanks for making the change (either formulation is fine).

**Comment by jpopelar** [comment]
@scheibelp PTAL when convenient, I think this is good to go after I rebase

**Review by scheibelp — DISMISSED** [review]

**Review by scheibelp — APPROVED** [review]

File: var/spack/repos/builtin/packages/openblas/package.py
Comment (johnwparent) [inline var/spack/repos/builtin/packages/openblas/package.py:277]: How does this work? Concatenating a string and a list like this results in the string being decomposed into chars and each char being added to the list, so at best this is a no op polluting the lib list?
@scheibelp

Comment (johnwparent) [reply var/spack/repos/builtin/packages/openblas/package.py:277]: Sorry to comment on a dead PR, I just went to look at this for the PATH issue when considering how to handle virtual deps and this jumped out at me.
Happy to submit a patch if I I'm not wrong about this. I've got one locally.

Comment (scheibelp) [reply var/spack/repos/builtin/packages/openblas/package.py:277]: > How does this work?

It almost definitely doesn't. A patch would be great.

Comment (johnwparent) [reply var/spack/repos/builtin/packages/openblas/package.py:277]: Ok, one will be up shortly!

**Pull request merged**
